From: Chad Monroe <chad@monroe.io>
Date: Tue, 17 Sep 2024 04:54:04 -0700
Subject: [PATCH] wifi: mt76: mt7996: fix single SKU power limits

* add mt7996 specific get_rate_power_limits() as the path power
  limit sizes are different from MT7915. there are better ways
  to solve this but not without significant modification to
  existing patches.
* add Canada to the list of PSD limited countries
* default sku_path_en to false and only enable it once a path
  power entry is found in the device tree
* re-apply single SKU limits after SER event
* convert mt7996 users of the 6 GHz LPI flags to per-phy storage
  introduced in the mt7915 counterpart of this patch. this
  also band-scopes the lpi_bcn_enhance beacon rate override in
  mt7996_get_rates_table() which otherwise forces OFDM 6M beacons
  on all bands

Signed-off-by: Chad Monroe <chad@monroe.io>
---
 mt7996/init.c   |    8 ++++---
 mt7996/mac.c    |    2 -
 mt7996/main.c   |   21 ++++++++++++++++--
 mt7996/mcu.c    |   20 ++++++++++++++---
 mt7996/vendor.c |   64 ++++++++++++++++++++++++++++++++------------------------
 mt7996/vendor.h |    2 -
 6 files changed, 79 insertions(+), 38 deletions(-)

--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -6283,7 +6283,7 @@ mt7996_update_max_txpower_cur(struct mt7
 
 static bool mt7996_is_psd_country(char *country)
 {
-	char psd_country_list[][3] = {"US", "KR", "BR", "CL", "MY", ""};
+	char psd_country_list[][3] = {"US", "CA", "KR", "BR", "CL", "MY", ""};
 	int i;
 
 	if (strlen(country) != 2)
@@ -6336,7 +6336,13 @@ int mt7996_mcu_set_txpower_sku(struct mt
 
 	if (txpower_setting == INT_MIN || txpower_setting > 127)
 		txpower_setting = 127;
+
+	dev_dbg(dev->mt76.dev, "%s: txpower_setting=%d phy->txpower=%d chainmask=0x%x n_chains=%d path_delta=%d",
+		__func__, txpower_setting, phy->txpower, mphy->chainmask, hweight16(mphy->chainmask),
+		mt76_tx_power_path_delta(hweight16(mphy->chainmask)));
+
 	txpower_limit = mt7996_get_power_bound(phy, txpower_setting);
+	//txpower_limit += mt76_tx_power_path_delta(hweight16(mphy->chainmask));
 
 	if (phy->sku_limit_en) {
 		la = kzalloc(sizeof(*la), GFP_KERNEL);
@@ -6351,6 +6357,9 @@ int mt7996_mcu_set_txpower_sku(struct mt
 		return 0;
 	}
 
+	dev_dbg(dev->mt76.dev, "%s: txpower_limit=%d ofdm[0]=%d mcs[0][0]=%d",
+		__func__, txpower_limit, la->ofdm[0], la->mcs[0][0]);
+
 	skb = mt76_mcu_msg_alloc(&dev->mt76, NULL,
 				 sizeof(req) + MT7996_SKU_PATH_NUM);
 	if (!skb) {
@@ -6364,8 +6373,8 @@ int mt7996_mcu_set_txpower_sku(struct mt
 	/* FW would compensate for PSD countries
 	 * driver doesn't need to do it
 	 */
-	if (phy->mt76->cap.has_6ghz && mphy->dev->lpi_psd &&
-	    !mt7996_is_psd_country(dev->mt76.alpha2)) {
+	if (phy->mt76->cap.has_6ghz && mphy->lpi_psd &&
+	    mt7996_is_psd_country(dev->mt76.alpha2)) {
 		switch (mphy->chandef.width) {
 		case NL80211_CHAN_WIDTH_20:
 			skb_put_data(skb, &la->eht[3], 8);
@@ -6432,7 +6441,7 @@ int mt7996_mcu_set_txpower_sku(struct mt
 	/* FW would NOT compensate in the case of BF backoff table
 	 * driver needs to compensate for LPI PSD
 	 */
-	if (phy->mt76->cap.has_6ghz && mphy->dev->lpi_psd) {
+	if (phy->mt76->cap.has_6ghz && mphy->lpi_psd) {
 		switch (mphy->chandef.width) {
 		case NL80211_CHAN_WIDTH_20:
 			skb_put_data(skb, &la->path.ru[5], 5);
@@ -6500,6 +6509,9 @@ int mt7996_mcu_set_lpi_psd(struct mt7996
 		.psd_limit = enable ? mt7996_is_psd_country(dev->mt76.alpha2) : 0,
 	};
 
+	dev_info(dev->mt76.dev, "%s: enable = %d, psd_limit = %d",
+			__func__, req.lpi_enable, req.psd_limit);
+
 	return mt76_mcu_send_msg(&phy->dev->mt76, MCU_WM_UNI_CMD(BAND_CONFIG),
 				 &req, sizeof(req), false);
 }
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -50,6 +50,19 @@ int mt7996_run(struct mt7996_phy *phy)
 	phy->sr_enable = true;
 	phy->enhanced_sr_enable = true;
 
+	/* needed to re-apply power tables after SER */
+	if (phy->mt76->chanctx) {
+		u8 delta;
+		int current_txpower;
+
+		delta = mt76_tx_power_path_delta(hweight16(phy->mt76->chainmask));
+		current_txpower = DIV_ROUND_UP(phy->mt76->txpower_cur + delta, 2);
+		ret = mt7996_mcu_set_txpower_sku(phy, current_txpower);
+	}
+
+	dev_info(dev->mt76.dev, "%s: band = %d, sku enable = %d, path enable = %d\n", __func__, phy->mt76->band_idx,
+		dev->dbg.sku_disable ? 0 : phy->sku_limit_en, dev->dbg.sku_disable ? 0 : phy->sku_path_en);
+
 	ret = mt7996_mcu_set_tx_power_ctrl(phy, UNI_TXPOWER_SKU_POWER_LIMIT_CTRL,
 						dev->dbg.sku_disable ? 0 : phy->sku_limit_en);
 
@@ -679,6 +692,8 @@ int mt7996_set_channel(struct mt76_phy *
 {
 	struct mt7996_phy *phy = mphy->priv;
 	int ret;
+	u8 delta;
+	int current_txpower;
 
 	if (mphy->offchannel)
 		mt7996_mac_update_beacons(phy);
@@ -697,7 +712,9 @@ int mt7996_set_channel(struct mt76_phy *
 	if (ret)
 		goto out;
 
-	ret = mt7996_mcu_set_txpower_sku(phy);
+	delta = mt76_tx_power_path_delta(hweight16(mphy->chainmask));
+	current_txpower = DIV_ROUND_UP(mphy->txpower_cur + delta, 2);
+	ret = mt7996_mcu_set_txpower_sku(phy, current_txpower);
 	if (ret)
 		goto out;
 
@@ -893,7 +910,7 @@ mt7996_get_rates_table(struct mt7996_phy
 	rate = mt76_connac2_mac_tx_rate_val(phy->mt76, conf, beacon, mcast);
 
 	if (beacon) {
-		if (dev->mt76.lpi_bcn_enhance)
+		if (phy->mt76->lpi_bcn_enhance)
 			rate = FR_RATE_IDX_OFDM_6M;
 
 		/* odd index for driver, even index for firmware */
--- a/mt7996/vendor.c
+++ b/mt7996/vendor.c
@@ -32,7 +32,7 @@ txpower_ctrl_policy[NUM_MTK_VENDOR_ATTRS
 	[MTK_VENDOR_ATTR_TXPOWER_CTRL_LPI_PSD] = { .type = NLA_U8 },
 	[MTK_VENDOR_ATTR_TXPOWER_CTRL_SKU_IDX] = { .type = NLA_U8 },
 	[MTK_VENDOR_ATTR_TXPOWER_CTRL_LPI_BCN_ENHANCE] = { .type = NLA_U8 },
-	[MTK_VENDOR_ATTR_TXPOWER_CTRL_LINK_ID] = { .type = NLA_U8 },
+	[MTK_VENDOR_ATTR_TXPOWER_CTRL_RADIO_IDX] = { .type = NLA_U8 },
 };
 
 static const struct nla_policy
@@ -103,6 +103,9 @@ static int mt7996_vendor_edcca_ctrl(stru
 
 	phy = dev->radio_phy[radio_idx];
 
+	dev_info(dev->mt76.dev, "%s: radio_idx=%u band=%u",
+			__func__, radio_idx, phy->mt76->band_idx);
+
 	switch (edcca_mode) {
 	case EDCCA_CTRL_SET_EN:
 		if (!tb[MTK_VENDOR_ATTR_EDCCA_CTRL_PRI20_VAL])
@@ -221,6 +224,10 @@ static int mt7996_vendor_pp_ctrl(struct
 
 	if (tb[MTK_VENDOR_ATTR_PP_MODE]) {
 		val8 = nla_get_u8(tb[MTK_VENDOR_ATTR_PP_MODE]);
+
+		dev_info(dev->mt76.dev, "%s: band_idx=%u band=%u pp_en=%u",
+				__func__, band_idx, phy->mt76->band_idx, val8);
+
 		switch (val8) {
 		case PP_DISABLE:
 		case PP_FW_MODE:
@@ -247,18 +254,14 @@ static int mt7996_vendor_txpower_ctrl(st
 				      int data_len)
 {
 #define FR_RATE_IDX_OFDM_6M 0x004b
-	struct mt7996_dev *dev;
+	struct ieee80211_hw *hw = wiphy_to_ieee80211_hw(wiphy);
+	struct mt7996_dev *dev = mt7996_hw_dev(hw);
 	struct mt7996_phy *phy;
 	struct mt76_phy *mphy;
-	struct ieee80211_vif *vif = wdev_to_ieee80211_vif(wdev);
-	struct mt7996_vif *mvif = (struct mt7996_vif *)vif->drv_priv;
-	struct mt7996_vif_link *mconf;
 	struct nlattr *tb[NUM_MTK_VENDOR_ATTRS_TXPOWER_CTRL];
-	struct mt76_power_limits la = {};
-	struct mt76_power_path_limits la_path = {};
-	u8 orig_sku_idx = mphy->sku_idx;
+	struct mt76_power_limits *la;
 	int err, current_txpower, delta;
-	u8 val, link_id = 0, idx;
+	u8 val, radio_idx = 0, idx;
 	s8 new_power_limit = 127;
 
 	err = nla_parse(tb, MTK_VENDOR_ATTR_TXPOWER_CTRL_MAX, data, data_len,
@@ -266,30 +269,23 @@ static int mt7996_vendor_txpower_ctrl(st
 	if (err)
 		return err;
 
-
-	if (ieee80211_vif_is_mld(vif) && tb[MTK_VENDOR_ATTR_TXPOWER_CTRL_LINK_ID]) {
-		link_id = nla_get_u8(tb[MTK_VENDOR_ATTR_TXPOWER_CTRL_LINK_ID]);
-
-		if (link_id >= IEEE80211_LINK_UNSPECIFIED)
-			return -EINVAL;
-	}
-
-	rcu_read_lock();
-	mconf = (struct mt7996_vif_link *)rcu_dereference(mvif->mt76.link[link_id]);
-	if (!mconf || !mconf->phy) {
-		rcu_read_unlock();
+	if (!tb[MTK_VENDOR_ATTR_TXPOWER_CTRL_RADIO_IDX])
 		return -EINVAL;
-	}
 
-	phy = mconf->phy;
-	rcu_read_unlock();
+	radio_idx = nla_get_u8(tb[MTK_VENDOR_ATTR_TXPOWER_CTRL_RADIO_IDX]);
+	if (!mt7996_radio_valid(dev, radio_idx))
+		return -EINVAL;
 
+	phy = dev->radio_phy[radio_idx];
 	mphy = phy->mt76;
 
 	if (mphy->cap.has_6ghz &&
 	    tb[MTK_VENDOR_ATTR_TXPOWER_CTRL_LPI_PSD]) {
 		val = nla_get_u8(tb[MTK_VENDOR_ATTR_TXPOWER_CTRL_LPI_PSD]);
-		mphy->dev->lpi_psd = val;
+		mphy->lpi_psd = val;
+
+		dev_info(mphy->dev->dev, "%s: band=%u set lpi_psd=%u",
+				__func__, phy->mt76->band_idx, val);
 
 		err = mt7996_mcu_set_lpi_psd(phy, val);
 		if (err)
@@ -297,11 +293,16 @@ static int mt7996_vendor_txpower_ctrl(st
 	}
 
 	if (tb[MTK_VENDOR_ATTR_TXPOWER_CTRL_SKU_IDX]) {
+		u8 orig_sku_idx = mphy->sku_idx;
 		struct ieee80211_channel *chan = mphy->main_chandef.chan;
 
 		if (!chan)
 			return -EINVAL;
 
+		la = kzalloc(sizeof(*la), GFP_KERNEL);
+		if (!la)
+			return -ENOMEM;
+
 		mphy->sku_idx = nla_get_u8(tb[MTK_VENDOR_ATTR_TXPOWER_CTRL_SKU_IDX]);
 
 		if (mt76_find_power_limits_node(mphy) == NULL)
@@ -311,12 +312,17 @@ static int mt7996_vendor_txpower_ctrl(st
 			mt7996_init_txpower(phy);
 		new_power_limit = chan->max_power;
 
+		dev_dbg(mphy->dev->dev, "%s: band=%u set sku_idx=%u",
+				__func__, phy->mt76->band_idx, mphy->sku_idx);
+
 		phy->sku_limit_en = true;
 		phy->sku_path_en = true;
 		mt76_get_rate_power_limits(mphy, chan, la, 127);
-		if (!la_path.ofdm[0])
+		if (!la->path.ofdm[0])
 			phy->sku_path_en = false;
 
+		kfree(la);
+
 		dev = phy->dev;
 		err = mt7996_mcu_set_tx_power_ctrl(phy, UNI_TXPOWER_SKU_POWER_LIMIT_CTRL,
 						   dev->dbg.sku_disable ? 0 : phy->sku_limit_en);
@@ -331,14 +337,18 @@ static int mt7996_vendor_txpower_ctrl(st
 	if (mphy->cap.has_6ghz &&
 	    tb[MTK_VENDOR_ATTR_TXPOWER_CTRL_LPI_BCN_ENHANCE]) {
 		val = nla_get_u8(tb[MTK_VENDOR_ATTR_TXPOWER_CTRL_LPI_BCN_ENHANCE]);
-		mphy->dev->lpi_bcn_enhance = val;
+		mphy->lpi_bcn_enhance = val;
 		idx = MT7996_BEACON_RATES_TBL + 2 * phy->mt76->band_idx;
 
+		dev_info(mphy->dev->dev, "%s: band=%u set lpi_bcn_enhance=%u",
+				__func__, phy->mt76->band_idx, val);
+
 		err = mt7996_mcu_set_fixed_rate_table(phy, idx, FR_RATE_IDX_OFDM_6M, true);
 		if (err)
 			return err;
 	}
 
+	delta = mt76_tx_power_path_delta(hweight16(phy->mt76->chainmask));
 	current_txpower = DIV_ROUND_UP(mphy->txpower_cur + delta, 2);
 	new_power_limit = min_t(s8, current_txpower, new_power_limit);
 
--- a/mt7996/vendor.h
+++ b/mt7996/vendor.h
@@ -16,7 +16,7 @@ enum mtk_vendor_attr_txpower_ctrl {
 	MTK_VENDOR_ATTR_TXPOWER_CTRL_LPI_PSD,
 	MTK_VENDOR_ATTR_TXPOWER_CTRL_SKU_IDX,
 	MTK_VENDOR_ATTR_TXPOWER_CTRL_LPI_BCN_ENHANCE,
-	MTK_VENDOR_ATTR_TXPOWER_CTRL_LINK_ID,
+	MTK_VENDOR_ATTR_TXPOWER_CTRL_RADIO_IDX,
 
 	/* keep last */
 	NUM_MTK_VENDOR_ATTRS_TXPOWER_CTRL,
--- a/mt7996/init.c
+++ b/mt7996/init.c
@@ -370,7 +370,7 @@ static void __mt7996_init_txpower(struct
 		return;
 
 	phy->sku_limit_en = true;
-	phy->sku_path_en = true;
+	phy->sku_path_en = false;
 	np = mt76_find_power_limits_node(phy->mt76);
 	for (i = 0; i < sband->n_channels; i++) {
 		struct ieee80211_channel *chan = &sband->channels[i];
@@ -380,8 +380,10 @@ static void __mt7996_init_txpower(struct
 		target_power = mt76_get_rate_power_limits(phy->mt76, chan,
 							  limits,
 							  target_power);
-		if (!limits->path.ofdm[0])
-			phy->sku_path_en = false;
+		phy->mt76->txpower_cur = max_t(int, phy->mt76->txpower_cur,
+							  target_power);
+		if (limits->path.ofdm[0])
+			phy->sku_path_en = true;
 
 		target_power += path_delta;
 		target_power = DIV_ROUND_UP(target_power, 2);
--- a/mt7996/mac.c
+++ b/mt7996/mac.c
@@ -919,7 +919,7 @@ void mt7996_mac_write_txwi(struct mt7996
 		if (mcast)
 			val |= MT_TXD6_DIS_MAT;
 		if (dev->mt76.phys[band_idx]->cap.has_6ghz &&
-		    dev->mt76.lpi_bcn_enhance &&
+		    dev->mt76.phys[band_idx]->lpi_bcn_enhance &&
 		    ieee80211_is_mgmt(hdr->frame_control))
 			val |= FIELD_PREP(MT_TXD6_BW, FW_CDBW_80MHZ);
 		txwi[6] |= cpu_to_le32(val);
